src/Tools/dotnet-monitor/CollectionRules/ExecuteAction.cs

```diff
@@ -0,0 +1,78 @@
+﻿using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options.Actions;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules
+{
+    internal sealed class ExecuteAction : ICollectionRuleAction<ExecuteOptions>
+    {
+        // Completion source that is signaled when the process exits
+        private TaskCompletionSource<int> _exitedSource;
+
+        /// <summary>
+        /// Gets a task that completes with the process exit code when it exits.
+        /// </summary>
+        public Task<int> ExitedTask => _exitedSource.Task;
+
+        public async Task<ActionResult> ExecuteAsync(ExecuteOptions options, DiagnosticsClient client, CancellationToken cancellationToken)
+        {
+            string path = options.Path;
+            string arguments = options.Arguments;
+
+            if (!File.Exists(path))
+            {
+                throw new FileNotFoundException("A file could not be found at the provided path: " + path);
+            }
+
+            // May want to capture stdout and stderr and return as part of the result in the future
+            Process process = new Process();
+
+            process.StartInfo = new ProcessStartInfo(path, arguments);
+            process.EnableRaisingEvents = true;
+
+            _exitedSource = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            if (!process.Start())
+            {
+                throw new InvalidOperationException($"Unable to start: {process.StartInfo.FileName} {process.StartInfo.Arguments}");
+            }
+
+            await WaitForExitAsync(process, cancellationToken).ConfigureAwait(false);
+
+            int exitCode = process.ExitCode;
+
+            if (!options.IgnoreExitCode && exitCode != 0)
+            {
+                // Do we want to use a specific type of Exception here? My thought was that the proper type of Exception may be dependent on the exitCode's value
```

Comment:
Never use `Exception` directly. Always use a derived exception. I think `InvalidOperationException` is fine.

Reply:
Did we want to have a more constrained set of exceptions that actions can return (similar to pipelines) so that the calling infrastructure can have an easier time handling failures?

Reply:
In general, I would say this is a good idea. However, if we open the actions up to extensibility, then we need to make that exception type available publicly as well. Otherwise, we end up with a mix of the prebuilt actions throwing the special exception and dotnet-monitor treating that as "yes, the action failed, but that's necessarily a bug" whereas external actions would only be able to throw exceptions that are not this type and all of them would be considered buggy behavior.